### DISCUSSION TITLE
Performance Issue - Pygame Support For Codon

### DISCUSSION THREAD

**Discussion by cool-dev-guy** [exaloop/codon#375] [Q&A]
I was working on using pygame with codon.
And i was successful in it,But :
```
for i in range(100000):
   pygame.draw.circle(surface, (255, 0, 0), (i, i), 10)
```
And it took:
```
-----------------------
|  Python  |   Codon  |
|   0.3s   |   0.9s   |
-----------------------
```
Here codon takes `more time`,is there a way i can do it in small time?
Run:`codon build -exe -release main.py`

**Comment by elisbyberi** [comment]
@cool-dev-guy Type conversion and function calls between Codon and CPython incurs a performance cost, which should be taken into account when designing and implementing code.

Since CPython has its own runtime environment, data must flow back and forth between CPython and Codon. 
This is not the case if you call a C function using `from C import func()`

By avoiding the lookup of the "circle" function in each iteration, you can optimize it in both CPython and Codon.

``` python
import time

from python import pygame  # Codon
# import pygame # CPython

pygame.init()

# Create a surface with size 640x480 pixels
surface = pygame.Surface((640, 480))

# time to draw circle using pygame.draw.circle
t = time.time()
for i in range(1000000):
    pygame.draw.circle(surface, (255, 0, 0), (i, i), 10)

print(time.time() - t)
# Codon: 1.95779 secs
# CPython: 0.70422 secs

# time to look up circle function
t = time.time()
for i in range(1000000):
    circle = pygame.draw.circle

print(time.time() - t)
# Codon: 0.477455 secs
# CPython: 0.14289 secs

# time to draw circle using circle
t = time.time()
circle = pygame.draw.circle
for i in range(1000000):
    circle(surface, (255, 0, 0), (i, i), 10)

print(time.time() - t)
# Codon: 1.1882 secs
# CPython: 0.62985 secs

```

**Reply by cool-dev-guy** [reply]
I was trying `from C import` but now i am facing an issue,My cpp code returns a `SDL_event` to codon,what data type would be added to recieve it?
```
from C import LIBRARY.func() ->cobj
x = func()
# i need to get the value returned not the pointer, How?
```

**Comment by cool-dev-guy** [comment]
1. I was trying `from C import` but now i am facing an issue,My cpp code returns a `SDL_event` to codon which is a struct data type,what data type would be added to recieve it in codon?

**Reply by elisbyberi** [reply]
@cool-dev-guy I created a minimal reproducible example (MRE): https://github.com/exaloop/codon/discussions/372#discussioncomment-5875595

**Comment by cool-dev-guy** [comment]
(Q1) is there any way to import cpp headers to codon (eg:SDL.h)? Or any other methods?

(Q2) does python's .so (made from setuptools) be directly imported to codon?(like : import pythonlib as pl)

**Reply by elisbyberi** [reply]
@cool-dev-guy C/C++ headers cannot be directly imported in Python or Codon.

You can import any Python module using `from python import module`: https://docs.exaloop.io/codon/interoperability/python
